Export fra API/Flights er begyndt at fejl efter den 24 september 2025 og tildeling af UnionId for nye er begyndt at fejle

Den automatiske export til DSvU har ikke fungeret siden slut september og det er fremfundet at der ved exporten sker en fejl i det lag som validerer hvilke flyvninger som skal medtages. Der er samtidig opstået meldinger på fejl ved opdatering af UnionsId på piloter som har fløjet og hvor der i koden er indbygget at tidligere flyvninger bliver opdateret således at de medtages i den kommende synkronisering til DSvU, denne kode fejler også nu. 

Der er ikke udført ændringer i koden de sidste 7 måneder i drift. 
Der er ikke udført ændringer til driftmiljøet i perioden som skulle have indvirkning på dette område. 

Kandidater til fejlkilder er logikken i selve entity framework hvor der mistænkes en n+1
Der er løbet data igennem og fundet enkelte pilot profiler med lang historik uden unionid men ikke nogen oplagt kandidater på data siden.

FlightJournal.Web\Controllers\PilotController.cs:166 Error code 8632.
A severe error occurred on the current command.  The results, if any, should be discarded.. har vedhæftet fuld stack trace nedenfor. https://learn.microsoft.com/en-us/sql/relational-databases/errors-events/mssqlserver-8632-database-engine-error?view=sql-server-ver17

Kald til API/Flights har fejlet i line 374 med System.Data.Entity.Core.EntityException
FlightJournal.Web\Controllers\TrainingLogHistoryController.cs:374

`[SqlException (0x80131904): The service has encountered an error processing your request. Please try again. Error code 8632.
A severe error occurred on the current command.  The results, if any, should be discarded.]
   System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction) +212
   System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction) +81
   System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose) +629
   System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady) +4412
   System.Data.SqlClient.SqlDataReader.TryConsumeMetaData() +58
   System.Data.SqlClient.SqlDataReader.get_MetaData() +89
   System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString, Boolean isInternal, Boolean forDescribeParameterEncryption, Boolean shouldCacheForAlwaysEncrypted) +438
   System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest) +2584
   System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry) +1700
   System.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method) +64
   System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior, String method) +977
   System.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior) +37
   System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior) +12
   System.Data.Entity.Infrastructure.Interception.<>c.<Reader>b__6_0(DbCommand t, DbCommandInterceptionContext`1 c) +20
   System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch(TTarget target, Func`3 operation, TInterceptionContext interceptionContext, Action`3 executing, Action`3 executed) +576
   System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.Reader(DbCommand command, DbCommandInterceptionContext interceptionContext) +430
   System.Data.Entity.Internal.InterceptableDbCommand.ExecuteDbDataReader(CommandBehavior behavior) +166
   System.Data.Common.DbCommand.ExecuteReader(CommandBehavior behavior) +12
   System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands(EntityCommand entityCommand, CommandBehavior behavior) +37

[EntityCommandExecutionException: An error occurred while executing the command definition. See the inner exception for details.]
   System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands(EntityCommand entityCommand, CommandBehavior behavior) +112
   System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlan.Execute(ObjectContext context, ObjectParameterCollection parameterValues) +744
   System.Data.Entity.Core.Objects.<>c__DisplayClass41_0.<GetResults>b__1() +97
   System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction(Func`1 func, IDbExecutionStrategy executionStrategy, Boolean startLocalTransaction, Boolean releaseConnectionOnSuccess) +292
   System.Data.Entity.Core.Objects.<>c__DisplayClass41_0.<GetResults>b__0() +178
   System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute(Func`1 operation) +75

[EntityException: An exception has been raised that is likely due to a transient failure. If you are connecting to a SQL Azure database consider using SqlAzureExecutionStrategy.]
   System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute(Func`1 operation) +185
   System.Data.Entity.Core.Objects.ObjectQuery`1.GetResults(Nullable`1 forMergeOption) +279
   System.Data.Entity.Core.Objects.ObjectQuery`1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__31_0() +16
   System.Data.Entity.Internal.LazyEnumerator`1.MoveNext() +45
   FlightJournal.Web.Controllers.PilotController.Edit(PilotDetailsViewModel vm) in D:\a\startlist.club\startlist.club\FlightJournal.Web\Controllers\PilotController.cs:166
   lambda_method(Closure , ControllerBase , Object[] ) +107
   System.Web.Mvc.ActionMethodDispatcher.Execute(ControllerBase controller, Object[] parameters) +19
   System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters) +169
   System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters) +27
   System.Web.Mvc.Async.<>c.<BeginInvokeSynchronousActionMethod>b__9_0(IAsyncResult asyncResult, ActionInvocation innerInvokeState) +24
   System.Web.Mvc.Async.WrappedAsyncResult`2.CallEndDelegate(IAsyncResult asyncResult) +31
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +50
   System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethod(IAsyncResult asyncResult) +33
   System.Web.Mvc.Async.<>c__DisplayClass11_0.<InvokeActionMethodFilterAsynchronouslyRecursive>b__0() +58
   System.Web.Mvc.Async.<>c__DisplayClass11_2.<InvokeActionMethodFilterAsynchronouslyRecursive>b__2() +228
   System.Web.Mvc.Async.<>c__DisplayClass7_0.<BeginInvokeActionMethodWithFilters>b__1(IAsyncResult asyncResult) +15
   System.Web.Mvc.Async.WrappedAsyncResult`1.CallEndDelegate(IAsyncResult asyncResult) +15
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +50
   System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeActionMethodWithFilters(IAsyncResult asyncResult) +35
   System.Web.Mvc.Async.<>c__DisplayClass3_5.<BeginInvokeAction>b__4() +32
   System.Web.Mvc.Async.<>c__DisplayClass3_0.<BeginInvokeAction>b__1(IAsyncResult asyncResult) +91
   System.Web.Mvc.Async.WrappedAsyncResult`1.CallEndDelegate(IAsyncResult asyncResult) +15
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +50
   System.Web.Mvc.Async.AsyncControllerActionInvoker.EndInvokeAction(IAsyncResult asyncResult) +29
   System.Web.Mvc.<>c.<BeginExecuteCore>b__152_1(IAsyncResult asyncResult, ExecuteCoreState innerState) +13
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +29
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +50
   System.Web.Mvc.Controller.EndExecuteCore(IAsyncResult asyncResult) +45
   System.Web.Mvc.<>c.<BeginExecute>b__151_2(IAsyncResult asyncResult, Controller controller) +15
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +22
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +50
   System.Web.Mvc.Controller.EndExecute(IAsyncResult asyncResult) +27
   System.Web.Mvc.Controller.System.Web.Mvc.Async.IAsyncController.EndExecute(IAsyncResult asyncResult) +12
   System.Web.Mvc.<>c.<BeginProcessRequest>b__20_1(IAsyncResult asyncResult, ProcessRequestState innerState) +28
   System.Web.Mvc.Async.WrappedAsyncVoid`1.CallEndDelegate(IAsyncResult asyncResult) +29
   System.Web.Mvc.Async.WrappedAsyncResultBase`1.End() +50
   System.Web.Mvc.MvcHandler.EndProcessRequest(IAsyncResult asyncResult) +29
   System.Web.Mvc.MvcHandler.System.Web.IHttpAsyncHandler.EndProcessRequest(IAsyncResult result) +11
   System.Web.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute() +574
   System.Web.<>c__DisplayClass285_0.<ExecuteStepImpl>b__0() +23
   System.Web.StepInvoker.Invoke(Action executionStep) +100
   System.Web.<>c__DisplayClass4_0.<Invoke>b__0() +22
   Microsoft.AspNet.TelemetryCorrelation.TelemetryCorrelationHttpModule.OnExecuteRequestStep(HttpContextBase context, Action step) +65
   System.Web.<>c__DisplayClass284_0.<OnExecuteRequestStep>b__0(Action nextStepAction) +54
   System.Web.StepInvoker.Invoke(Action executionStep) +84
   System.Web.<>c__DisplayClass4_0.<Invoke>b__0() +22
   Microsoft.ApplicationInsights.Web.ApplicationInsightsHttpModule.OnExecuteRequestStep(HttpContextBase context, Action step) in /_/WEB/Src/Web/Web/ApplicationInsightsHttpModule.cs:164
   System.Web.<>c__DisplayClass284_0.<OnExecuteRequestStep>b__0(Action nextStepAction) +54
   System.Web.StepInvoker.Invoke(Action executionStep) +84
   System.Web.HttpApplication.ExecuteStepImpl(IExecutionStep step) +80
   System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously) +163`


`System.Data.Entity.Core.EntityException:
   at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute (EntityFramework.SqlServer, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.GetResults (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.Objects.ObjectQuery`1.<System.Collections.Generic.IEnumerable<T>.GetEnumerator>b__31_0 (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Internal.LazyEnumerator`1.MoveNext (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at FlightJournal.Web.Controllers.TrainingLogHistoryController.CreateExportModel (FlightJournal.Web, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: D:\a\startlist.club\startlist.club\FlightJournal.Web\Controllers\TrainingLogHistoryController.cs:374)
   at FlightJournal.Web.Controllers.api.FlightsController.GetFlights (FlightJournal.Web, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: D:\a\startlist.club\startlist.club\FlightJournal.Web\Controllers\api\FlightsController.cs:59)
   at lambda_method (Anonymously Hosted DynamicMethods Assembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor+ActionExecutor+<>c__DisplayClass10.<GetExecutor>b__9 (System.Web.Http, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor+ActionExecutor.Execute (System.Web.Http, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Web.Http.Controllers.ReflectedHttpActionDescriptor.ExecuteAsync (System.Web.Http, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Web.Http.Controllers.ApiControllerActionInvoker+<InvokeActionAsyncCore>d__0.MoveNext (System.Web.Http, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Web.Http.Controllers.ActionFilterResult+<ExecuteAsync>d__2.MoveNext (System.Web.Http, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Web.Http.Dispatcher.HttpControllerDispatcher+<SendAsync>d__1.MoveNext (System.Web.Http, Version=5.2.3.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35)
Inner exception System.Data.Entity.Core.EntityCommandExecutionException handled at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute:
   at System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.Objects.Internal.ObjectQueryExecutionPlan.Execute (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.Objects.ObjectQuery`1+<>c__DisplayClass41_0.<GetResults>b__1 (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.Objects.ObjectContext.ExecuteInTransaction (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.Objects.ObjectQuery`1+<>c__DisplayClass41_0.<GetResults>b__0 (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.SqlServer.DefaultSqlExecutionStrategy.Execute (EntityFramework.SqlServer, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
Inner exception System.Data.SqlClient.SqlException handled at System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands:
   at System.Data.SqlClient.SqlConnection.OnError (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlInternalConnection.OnError (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.TdsParser.TryRun (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlDataReader.get_MetaData (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlCommand.RunExecuteReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlCommand.ExecuteReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.SqlClient.SqlCommand.ExecuteDbDataReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Common.DbCommand.ExecuteReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher+<>c.<Reader>b__6_0 (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Infrastructure.Interception.InternalDispatcher`1.Dispatch (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Infrastructure.Interception.DbCommandDispatcher.Reader (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Internal.InterceptableDbCommand.ExecuteDbDataReader (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Common.DbCommand.ExecuteReader (System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Data.Entity.Core.EntityClient.Internal.EntityCommandDefinition.ExecuteStoreCommands (EntityFramework, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)`

--- a/FlightJournal.Web/App_Start/DbCommandInterceptor.cs
+++ b/FlightJournal.Web/App_Start/DbCommandInterceptor.cs
@@ -73,11 +73,20 @@ namespace FlightJournal.Web.App_Start
 #endif
             if (interceptionContext.Exception != null)
             {
-                System.Diagnostics.Trace.TraceError($"Command {command.CommandText} failed with exception {interceptionContext.Exception}");
+                // Check for the specific error about 'CreatedOn' in '__MigrationHistory' that is a check done by EF between 4.3 and 5 on init of databases
+                if (command.CommandText.Contains("SELECT TOP (1) \r\n    [c].[CreatedOn] AS [CreatedOn]\r\n    FROM [dbo].[__MigrationHistory] AS [c]"))
+                {
+                    // Ignore this specific exception
+                    // EF __Migration exception with CreatedOn is related to https://stackoverflow.com/a/20670134 and is being throwned often but captured 
+                }
+                else
+                {
+                    System.Diagnostics.Trace.TraceError($"Command {command.CommandText} failed with exception {interceptionContext.Exception}");
 #if (DEBUG)
-                // Note that EF __Migration exception with CreatedOn is related to https://stackoverflow.com/a/20670134 and is being throwned often but captured 
-                throw new Exception($"Command {command.CommandText} failed with exception {interceptionContext.Exception}");
+                    throw new Exception($"Command {command.CommandText} failed with exception {interceptionContext.Exception}");
 #endif
+                }
+
             }
             else
             {

--- a/FlightJournal.Web/Controllers/PilotController.cs
+++ b/FlightJournal.Web/Controllers/PilotController.cs
@@ -153,22 +153,29 @@ namespace FlightJournal.Web.Controllers
                 if (pilotInDb.UnionId != vm.UnionId)
                 {
                     // mark all (training) flights where the pilot participated as changed - will ensure re-export to FA
-                    var affectedFlights = db.Flights.Where(f =>
-                        f.Deleted == null 
-                        && (f.PilotId == vm.PilotId 
-                            || (f.PilotBackseatId.HasValue && f.PilotBackseatId.Value == vm.PilotId)));
-                    // paranoia check, HasTrainingData was introduced during 2021  (TODO: script a DB update) - note that this has still been observed, apparently a quick user can still manage to not set HastrainingData.
+                    var affectedFlights = db.Flights
+                        .Where(f => f.Deleted == null &&
+                                    (f.PilotId == vm.PilotId ||
+                                        (f.PilotBackseatId.HasValue && f.PilotBackseatId.Value == vm.PilotId)))
+                        .ToList(); // Materialize the query to avoid multiple enumerations
+
+                    // paranoia check, HasTrainingData was introduced during 2021
                     var trainingFlightIds = db.AppliedExercises
                         .Select(x => x.FlightId)
                         .Distinct()
                         .ToHashSet();
-                    affectedFlights = affectedFlights.Where(x => x.HasTrainingData || trainingFlightIds.Contains(x.FlightId));
-                    foreach (var f in affectedFlights)
+
+                    // Filter affected flights based on HasTrainingData or trainingFlightIds
+                    var validAffectedFlights = affectedFlights
+                        .Where(x => x.HasTrainingData || trainingFlightIds.Contains(x.FlightId))
+                        .ToList(); // Materialize the filtered results
+
+                    foreach (var f in validAffectedFlights)
                     {
                         db.Entry(f).State = EntityState.Modified;
                         f.LastUpdated = DateTime.Now;
                         f.LastUpdatedBy = User.Pilot().ToString();
-                        // don't notify - that would probably create too much noise
+                        // don't notify - would create too much noise when done in a mass edit
                         // FlightsHub.NotifyFlightChanged(f.FlightId, Guid.Empty);
                     }
                 }

--- a/FlightJournal.Web/Controllers/api/FlightsController.cs
+++ b/FlightJournal.Web/Controllers/api/FlightsController.cs
@@ -45,6 +45,7 @@ namespace FlightJournal.Web.Controllers.api
             if (unionId != null)
                 flights = flights.Where(f => f.Pilot.UnionId == unionId);
 
+            var affectedFlights = flights.ToList(); // Materialize the query to avoid multiple enumerations
 
             if (training)
             {
@@ -53,11 +54,13 @@ namespace FlightJournal.Web.Controllers.api
                     .Select(x => x.FlightId)
                     .Distinct()
                     .ToHashSet();
-                flights = flights.Where(x => x.HasTrainingData || trainingFlightIds.Contains(x.FlightId));
+                
+                affectedFlights = affectedFlights
+                        .Where(x => x.HasTrainingData || trainingFlightIds.Contains(x.FlightId))
+                        .ToList(); // Materialize the filtered results
             }
 
-            var models = TrainingLogHistoryController.CreateExportModel(_db, flights, "");
-
+            var models = TrainingLogHistoryController.CreateExportModel(_db, affectedFlights, "");
             return Json(JsonConvert.SerializeObject(models, Formatting.None));
         }
 
@@ -66,7 +69,35 @@ namespace FlightJournal.Web.Controllers.api
             public string t;
             public string id;
         }
+
+        [HttpGet]
+        [AllowAnonymous]
+        [Route("query")]
+        public IHttpActionResult GetApiQuery()
+        {
+            var html = @"
+        <html>
+            <head><title>Flights API Test Query</title></head>
+            <body>
+                <h1>Flights API Test</h1>
+                <form method='get' action='/api/flights'>
+                    <label>API Key: <input name='key' /></label><br/>
+                    <label>From (yyyyMMdd): <input name='from' /></label><br/>
+                    <label>To (yyyyMMdd): <input name='to' /></label><br/>
+                    <label>Updated Since (yyyyMMdd): <input name='updatedSince' /></label><br/>
+                    <label>Training: <input type='checkbox' name='training' value='true' /></label><br/>
+                    <label>Pilot ID: <input name='pilotId' type='number' /></label><br/>
+                    <label>Union ID: <input name='unionId' /></label><br/>
+                    <input type='submit' value='Check API' />
+                </form>
+                <p>Results will be shown as JSON.</p>
+            </body>
+        </html>";
+            return new System.Web.Http.Results.ResponseMessageResult(
+                new System.Net.Http.HttpResponseMessage(System.Net.HttpStatusCode.OK)
+                {
+                    Content = new System.Net.Http.StringContent(html, System.Text.Encoding.UTF8, "text/html")
+                });
+        }
     }
-
-
 }

